icu4c/source/i18n/measunit_extra.cpp

```diff
@@ -194,6 +194,65 @@ class SimpleUnitIdentifiersSink : public icu::ResourceSink {
     int32_t outIndex;
 };
 
+/**
+ * A ResourceSink that collects information from `unitQuantities` in the `units`
+ * resource (see `units.txt`). unitQuantity values are collected into a char*
+ * array, while unitQuantity keys are added added to a TrieBuilder, with
+ * associated values being the index into the aforementioned char* array.
+ */
+class UnitQuantitiesSink : public icu::ResourceSink {
+  public:
+    /**
+     * Constructor.
+     * @param out Array of char* to which unitQuantity values will be written.
+     * @param outSize The size of the `out` array.
+     * @param trieBuilder The trie builder to which the keys (base units) of
+     *     each unitQuantity will be added, each with value being the offset
+     *     into `out`.
+     */
+    explicit UnitQuantitiesSink(char **out, int32_t &outSize, BytesTrieBuilder &trieBuilder)
+        : outQuantitiesArray(out), outSize(outSize), trieBuilder(trieBuilder), outIndex(0) {}
+
+    void put(const char * /*key*/, ResourceValue &value, UBool /*noFallback*/, UErrorCode &status) {
+        ResourceArray array = value.getArray(status);
+        if (U_FAILURE(status)) return;
+
+        if (outIndex + array.getSize() > outSize) {
+            status = U_INDEX_OUTOFBOUNDS_ERROR;
+            return;
+        }
+
+        for (int32_t i = 0; array.getValue(i, value); ++i) {
+            U_ASSERT(i < array.getSize());
+            U_ASSERT(outIndex < outSize);
+            ResourceTable table = value.getTable(status);
+            U_ASSERT(table.getSize() == 1);
```

Comment:
It is obvious that the table size always equal to 1. so, I do not know why do we need it

Reply:
Clear how, where? By looking at units.txt? Data might change, so code assumptions might break, and this is also intended as documentation to the reader of the code: we're only calling getKeyAndValue with index 0. "Why not other indices?" the reader might wonder, "ah, because there is only one, as is clearly asserted". Does that make more sense?

Reply:
+1 to what Hugo said.  Technically setting an error code is better than an assertion though, since ICU should be at least *somewhat* fault-tolerant to bad data files.  I don't remember the error we normally use in this situation; search the code base for examples.

Reply:
Going with U_INVALID_FORMAT_ERROR, looks reasonable enough.